pkg/apis/batch/validation/validation.go

```diff
@@ -88,6 +88,14 @@ func ValidateJob(job *batch.Job, opts apivalidation.PodValidationOptions) field.
 	allErrs := apivalidation.ValidateObjectMeta(&job.ObjectMeta, true, apivalidation.ValidateReplicationControllerName, field.NewPath("metadata"))
 	allErrs = append(allErrs, ValidateGeneratedSelector(job)...)
 	allErrs = append(allErrs, ValidateJobSpec(&job.Spec, field.NewPath("spec"), opts)...)
+	if job.Spec.CompletionMode != nil && *job.Spec.CompletionMode == batch.IndexedCompletion && len(job.ObjectMeta.Name) > apimachineryvalidation.DNS1035LabelMaxLength-11 {
```

Comment:
we add one more dash to the `generateName`: `$(job-name)-$(index)-`. So maybe 12 is a better number.
However, I'm not sure what happens if `generateName` already has 63 characters. @liggitt ?

Reply:
It looks like apiserver handles the truncation of generateName to make space for the random characters.

/lgtm

Leaving the unhold to Jordan.